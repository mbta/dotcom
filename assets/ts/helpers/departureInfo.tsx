import React from "react";
import {
  chain,
  concat,
  difference,
  find,
  keyBy,
  keys,
  map,
  sortBy
} from "lodash";
import { PredictionWithTimestamp } from "../models/predictions";
import { ScheduleWithTimestamp } from "../models/schedules";
import { isCancelled, isDelayed, isSkipped } from "./prediction-helpers";
import { DepartureInfo } from "../models/departureInfo";
import {
  isABusRoute,
  isACommuterRailRoute,
  isSubwayRoute
} from "../models/route";
import { DirectionId, Route } from "../__v3api";
import DisplayTime from "../stop/components/DisplayTime";
import { getInfoKey } from "../stop/models/displayTimeConfig";
import { RoutePatternWithPolyline } from "../models/route-patterns";

export const SUBWAY = "subway";
export const BUS = "bus";
export const COMMUTER_RAIL = "commuter_rail";
export const FERRY = "ferry";

const toRouteMode = (
  route: Route
): typeof SUBWAY | typeof BUS | typeof COMMUTER_RAIL | typeof FERRY => {
  if (isSubwayRoute(route)) {
    return SUBWAY;
  }
  if (isABusRoute(route)) {
    return BUS;
  }
  if (isACommuterRailRoute(route)) {
    return COMMUTER_RAIL;
  }
  return FERRY;
};

const departureInfoToTime = (
  departureInfo: DepartureInfo
): Date | undefined => {
  return departureInfo.prediction?.time || departureInfo.schedule?.time;
};

const getNextUnCancelledDeparture = (
  departureInfos: DepartureInfo[]
): DepartureInfo | undefined => {
  return find(departureInfos, (di: DepartureInfo) => !di.isCancelled);
};

const displayInfoContainsPrediction = (
  departureInfo: DepartureInfo
): boolean => {
  return !!departureInfo.prediction;
};

const mergeIntoDepartureInfo = (
  schedules: ScheduleWithTimestamp[],
  predictions: PredictionWithTimestamp[]
): DepartureInfo[] => {
  const schedulesByTripId = keyBy(schedules, "trip.id");
  const predictionsByTripId = keyBy(predictions, "trip.id");
  const scheduleTripIds = keys(schedulesByTripId);
  const predictionTripIds = keys(predictionsByTripId);

  // All the schedules with and without predictions
  const scheduleDirectionInfo = map(
    schedulesByTripId,
    (schedule, tripId): DepartureInfo => {
      // prediction could be undefined
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        schedule,
        route: schedule.route,
        trip: schedule.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        isDelayed: isDelayed(prediction, schedule),
        routeMode: toRouteMode(schedule.route)
      };
    }
  );

  const unmatchedPredictionTripIds = difference(
    predictionTripIds,
    scheduleTripIds
  );
  // All the predictions with no schedules
  const predictionDirectionInfo = map(
    unmatchedPredictionTripIds,
    (tripId: string): DepartureInfo => {
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        route: prediction.route,
        trip: prediction.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        routeMode: toRouteMode(prediction.route)
      } as DepartureInfo;
    }
  );

  const departureInfos = concat(scheduleDirectionInfo, predictionDirectionInfo);
  const sortedDepartureInfo = sortBy(departureInfos, di => {
    // prioritize sorting by predictions
    if (di.prediction && di.prediction.time) {
      return di.prediction.time;
    }
    return di.schedule?.time;
  });

  return sortedDepartureInfo;
};

const isAtDestination = (
  stopName: string,
  route: Route,
  directionId: DirectionId
): boolean => {
  const destination = route.direction_destinations[directionId];
  if (destination) {
    return destination.includes(stopName) && !destination.includes(" or ");
  }
  return false;
};

const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
  return departureInfo?.routeMode === "commuter_rail";
};

const DefaultWrapper: React.FunctionComponent<unknown> = ({ children }) => (
  <li>{children}</li>
);
/**
 * From a list of `DepartureInfo[]`, generate a list of `<DisplayTime />`.
 * Each `<DisplayTime />` will recieve the `isCR` and `targetDate` props.
 *
 * This function can optionally
 * - wrap each `<DisplayTime />` in a specified wrapper element, `WrapperEl`
 * - truncate the list to a desired `listLength` value
 */
const departuresListFromInfos = (
  departureInfos: DepartureInfo[],
  isCR: boolean,
  isSubway: boolean,
  targetDate?: Date,
  listLength?: number,
  omitCancelledAndSkipped = false,
  WrapperEl: typeof DefaultWrapper = DefaultWrapper
): React.ReactElement[] => {
  const predictions = chain(departureInfos)
    .reject(
      departure =>
        (omitCancelledAndSkipped &&
          (!!departure.isCancelled || !!departure.isSkipped)) ||
        (isSubway && typeof departure.prediction === "undefined")
    )
    .slice(0, listLength)
    .value();

  if (predictions.length === 0) {
    const routeId = departureInfos[0]?.route?.id;
    const tripId = departureInfos[0]?.trip?.id;

    return [
      <div className="text-sm font-helvetica" key={`${routeId}-${tripId}`}>
        No real-time data
      </div>
    ];
  }

  return predictions.map(d => (
    <WrapperEl key={getInfoKey(d)}>
      <DisplayTime departure={d} isCR={isCR} targetDate={targetDate} />
    </WrapperEl>
  ));
};

const departureInfoInRoutePatterns = (
  departureInfo: DepartureInfo,
  routePatterns: RoutePatternWithPolyline[]
): boolean => {
  const isMatch = !!routePatterns.find(
    rp => rp.id === departureInfo.trip.route_pattern_id
  );
  // This second case is needed for the edge case where the route pattern is
  // canonical but the schedules/predictions are not serving canonical route
  // patterns (e.g. due to shuttles or other disruptions)
  const isSimilar = !!routePatterns.find(
    rp =>
      rp.headsign === departureInfo.trip.headsign &&
      rp.route_id === departureInfo.route.id
  );
  return isMatch || isSimilar;
};

export {
  mergeIntoDepartureInfo,
  departureInfoToTime,
  displayInfoContainsPrediction,
  getNextUnCancelledDeparture,
  isAtDestination,
  isCommuterRail,
  departuresListFromInfos,
  departureInfoInRoutePatterns
};
