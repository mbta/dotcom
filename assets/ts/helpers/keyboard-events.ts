export const isEnter = (key: number | string): boolean =>
  key === "Enter" || key === 13;

export const isEscape = (key: number | string): boolean =>
  key === "Escape" || key === 27;

// for vanilla JS
export const handleNativeEnterKeyPress = (
  e: KeyboardEvent,
  cb: Function
): void => (isEnter(e.key || e.keyCode) ? cb(e) : (() => {})());

export const handleNativeEscapeKeyPress = (
  e: KeyboardEvent,
  cb: Function
): void => (isEscape(e.key || e.keyCode) ? cb(e) : (() => {})());
