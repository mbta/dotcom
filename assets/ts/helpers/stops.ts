import { Stop } from "../__v3api";
// eslint-disable-next-line import/prefer-default-export
export const isStopAStation = (stop: Stop): boolean => stop["station?"];

export const isStopAFerryStop = (stop: Stop): boolean => stop["ferry?"];
