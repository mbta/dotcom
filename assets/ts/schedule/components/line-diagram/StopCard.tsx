import React, { ReactElement, useContext } from "react";
import { effectNameForAlert } from "../../../components/Alerts";
import MatchHighlight from "../../../components/MatchHighlight";
import { alertIcon } from "../../../helpers/icon";
import {
  hasBranches,
  isEndNode,
  isStartNode,
  stopForId
} from "../../../helpers/stop-tree";
import {
  alertsByStop,
  isActiveDiversion,
  isCurrentLifecycle,
  isHighSeverityOrHighPriority
} from "../../../models/alert";
import { hasPredictionTime } from "../../../models/prediction";
import {
  isACommuterRailRoute,
  isAGreenLineRoute,
  isSubwayRoute
} from "../../../models/route";
import { Alert, Route } from "../../../__v3api";
import { RouteStop, StopId, StopTree } from "../__schedule";
import { branchPosition, diagramWidth } from "./line-diagram-helpers";
import StopConnections from "./StopConnections";
import StopFeatures from "./StopFeatures";
import StopPredictions from "./StopPredictions";
import { StopRefContext } from "./LineDiagramWithStops";
import { LiveData } from "./__line-diagram";

interface Props {
  alerts: Alert[];
  liveData?: LiveData;
  onClick: (stop: RouteStop) => void;
  routeStopList: RouteStop[];
  searchQuery?: string;
  stopId: StopId;
  stopTree: StopTree | null;
}

const width = (stopTree: StopTree, stopId: StopId): number =>
  diagramWidth(branchPosition(stopTree, stopId));

const hasBranchLabel = (stopTree: StopTree, stopId: StopId): boolean => {
  const stop: RouteStop = stopForId(stopTree, stopId);
  return (
    hasBranches(stopTree) &&
    (isStartNode(stopTree, stopId) || isEndNode(stopTree, stopId)) &&
    !!stop.branch &&
    !!stop.route
  );
};

const lineName = ({ name, route: routeStopRoute }: RouteStop): string => {
  const route = routeStopRoute as Route;
  const title = isAGreenLineRoute(route)
    ? `Green Line ${route.id.split("-")[1]}`
    : name;
  const lineOrBranch = isACommuterRailRoute(route) ? "Line" : "Branch";
  return `${title} ${lineOrBranch}`;
};

const hasLivePredictions = (liveData?: LiveData): boolean =>
  !!liveData && liveData.headsigns.some(hasPredictionTime);

const connectionsFor = (routeStop: RouteStop, stopTree: StopTree): Route[] => {
  const { connections } = routeStop;
  const greenLineConnections = connections.filter(isAGreenLineRoute);
  if (routeStop.route && hasBranches(stopTree) && greenLineConnections.length) {
    // If we can connect to other Green Line routes, they can connect back to
    // this route as well.
    return [routeStop.route, ...connections];
  }
  return connections;
};

const visibleAlert = (alert: Alert): boolean => {
  return isHighSeverityOrHighPriority(alert) && isCurrentLifecycle(alert);
};

const hasHighPriorityAlert = (stopId: StopId, alerts: Alert[]): boolean =>
  alertsByStop(alerts, stopId).filter(visibleAlert).length > 0;

const routeForStop = (stopTree: StopTree, stopId: StopId): Route | null => {
  const { route } = stopForId(stopTree, stopId);
  return route;
};

const hasUpcomingDeparturesIfSubway = (
  stopTree: StopTree,
  stopId: StopId,
  liveData?: LiveData
): boolean => {
  const route = routeForStop(stopTree, stopId);
  if (!route || !isSubwayRoute(route)) return true;
  return !!liveData && liveData.headsigns.length > 0;
};

const schedulesButtonLabel = (route: Route | null): string => {
  return route && isSubwayRoute(route)
    ? "View upcoming departures"
    : "View schedule";
};

const Alert = (): JSX.Element => (
  <>
    {alertIcon("c-svg__icon-alerts-triangle")}
    <span className="sr-only">Service alert or delay</span>
    &nbsp;
  </>
);

const StopCard = ({
  alerts,
  liveData,
  onClick,
  routeStopList,
  searchQuery,
  stopId,
  stopTree
}: Props): ReactElement<HTMLElement> => {
  const refs = useContext(StopRefContext)[0];
  const routeStop = stopTree
    ? stopForId(stopTree, stopId)
    : routeStopList.find(rs => rs.id === stopId)!;
  const routeStopIndex = routeStopList.indexOf(routeStop);
  const isEnd = stopTree
    ? isEndNode(stopTree, stopId)
    : routeStopIndex === routeStopList.length - 1;

  const diversionAlert = alerts.find(isActiveDiversion);
  const showDiversion =
    diversionAlert && !(hasLivePredictions(liveData) && isEnd);

  const left = stopTree ? width(stopTree, stopId) : diagramWidth(1);
  const connections = stopTree
    ? connectionsFor(routeStop, stopTree)
    : routeStop.connections;

  return (
    <li
      className="m-schedule-diagram__stop"
      style={{
        paddingLeft: searchQuery ? "0.5rem" : `${left}px`
      }}
    >
      <section className="m-schedule-diagram__content">
        {stopTree && hasBranchLabel(stopTree, stopId) && (
          <div className="font-bold u-small-caps">{lineName(routeStop)}</div>
        )}
        <header
          className="m-schedule-diagram__stop-heading"
          ref={el => refs.set(stopId, el)}
        >
          <h4 className="m-schedule-diagram__stop-link notranslate">
            <a href={`/stops/${stopId}`}>
              {hasHighPriorityAlert(stopId, alerts) && <Alert />}
              <MatchHighlight text={routeStop.name} matchQuery={searchQuery} />
            </a>
          </h4>
          {StopFeatures(routeStop)}
        </header>

        <div className="m-schedule-diagram__stop-details">
          {StopConnections(stopId, connections)}
          {hasLivePredictions(liveData) && !isEnd ? (
            <StopPredictions
              headsigns={liveData!.headsigns}
              isCommuterRail={
                !!routeStop.route && isACommuterRailRoute(routeStop.route)
              }
            />
          ) : (
            showDiversion && (
              <div className="m-schedule-diagram__alert">
                {effectNameForAlert(diversionAlert!)}
              </div>
            )
          )}
        </div>

        {!isEnd &&
          (stopTree
            ? hasUpcomingDeparturesIfSubway(stopTree, stopId, liveData)
            : true) && (
            <footer className="m-schedule-diagram__footer">
              <button
                className="btn btn-link"
                type="button"
                onClick={() => onClick(routeStop)}
              >
                {schedulesButtonLabel(
                  stopTree ? routeForStop(stopTree, stopId) : routeStop.route
                )}
              </button>
            </footer>
          )}
      </section>
    </li>
  );
};

export default StopCard;
