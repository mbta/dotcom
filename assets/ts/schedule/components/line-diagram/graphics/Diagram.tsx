import React, { ReactElement } from "react";
import { routeToModeName } from "../../../../helpers/css";
import {
  isBranchingNode,
  largestSliceSize,
  longestPath,
  longestPathLength,
  nextStopIds,
  stopForId,
  stopIds
} from "../../../../helpers/stop-tree";
import { isAGreenLineRoute } from "../../../../models/route";
import { Alert, DirectionId, Route } from "../../../../__v3api";
import {
  IndexedRouteStop,
  RouteStop,
  StopId,
  StopTree
} from "../../__schedule";
import { diagramWidth } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";
import Stop from "./Stop";
import { Line, SimpleLine } from "./Line";
import Merges from "./Merges";

interface Props {
  stopTree: StopTree;
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const routeName = (route: Route): string =>
  isAGreenLineRoute(route) ? "Green Line" : route.name;

const diagramDescription = (
  pathLength: number,
  route: Route,
  directionId: DirectionId
): string => {
  const text = `${pathLength} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = route;
  if (destinations) {
    return `${text} going ${names[directionId]} to ${destinations[directionId]}`;
  }
  return text;
};

const hasBranchLines = (stopTree: StopTree): boolean =>
  largestSliceSize(stopTree) > 1;

const branchingDescription = (stopTree: StopTree): string => {
  const primaryPath = longestPath(stopTree);
  const branchingStops: StopId[] = primaryPath.filter(id =>
    isBranchingNode(stopTree, id)
  );
  const branchText = branchingStops.map(branchingStopId => {
    const routeStop: RouteStop = stopForId(stopTree, branchingStopId);
    const branchNames: string[] = nextStopIds(stopTree, branchingStopId)
      .filter(id => !primaryPath.includes(id))
      .map(id => stopForId(stopTree, id)?.branch || "");

    return `Use ${routeStop.name} to transfer to trains on ${branchNames.join(
      " or "
    )} branches.`;
  });
  return branchText.join("");
};

interface SimpleProps {
  routeStopList: IndexedRouteStop[];
  route: Route;
  directionId: DirectionId;
  alerts: Alert[];
}

const SimpleDiagram = ({
  routeStopList,
  route,
  directionId,
  alerts
}: SimpleProps): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(routeStopList.length, route, directionId)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>
      {/* Draw lines between stops */
      routeStopList.map((routeStop, index) => {
        const nextStop = routeStopList[index + 1];
        if (!nextStop) return null;
        return (
          <SimpleLine
            key={`${routeStop.id}-${nextStop.id}`}
            fromId={routeStop.id}
            toId={nextStop.id}
            alerts={alerts}
          />
        );
      })}

      {/* Draw circles for each stop */
      routeStopList.map(routeStop => (
        <Stop key={`stop-${routeStop.id}`} stopId={routeStop.id} />
      ))}
    </svg>
  </>
);

const Diagram = ({
  stopTree,
  route,
  directionId,
  alerts
}: Props): ReactElement<HTMLElement> => (
  <>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      role="img"
      aria-labelledby="diagram-title diagram-desc"
      className={`line-diagram-svg ${routeToModeName(route)}`}
      width={`${diagramWidth(1)}px`}
      height="100%"
      style={{ left: BASE_LINE_WIDTH / 2, overflow: "visible" }}
    >
      <title id="diagram-title">Line diagram for {routeName(route)}</title>
      <desc id="diagram-desc">
        {diagramDescription(longestPathLength(stopTree), route, directionId)}
        {hasBranchLines(stopTree) && branchingDescription(stopTree)}
      </desc>
      <defs>{DiagonalHatchPattern()}</defs>

      {/* If there are multiple branches, draw the lines and curves for those */
      hasBranchLines(stopTree) && (
        <Merges stopTree={stopTree} alerts={alerts} />
      )}

      {/* Draw lines between stops */
      stopIds(stopTree).map(stopId =>
        nextStopIds(stopTree, stopId).map(nextStopId => (
          <Line
            key={`${stopId}-${nextStopId}`}
            stopTree={stopTree}
            fromId={stopId}
            toId={nextStopId}
            alerts={alerts}
          />
        ))
      )}

      {/* Draw circles for each stop */
      stopIds(stopTree).map(stopId => (
        <Stop key={stopId} stopId={stopId} />
      ))}
    </svg>
  </>
);

export { Diagram, SimpleDiagram };
