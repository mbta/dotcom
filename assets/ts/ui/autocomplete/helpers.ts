import { AutocompleteRenderer } from "@algolia/autocomplete-js";
import { omit } from "lodash";
import { createElement, Fragment } from "react";
import { render } from "react-dom";
import { parseQuery } from "../../helpers/query";
import {
  AutocompleteItem,
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  SearchResultItem,
  StopItem
} from "./__autocomplete";

export function isAlgoliaItem(x: Item): x is AutocompleteItem {
  return Object.keys(x).includes("objectID");
}

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export function isSearchResultItem(x: Item): x is SearchResultItem {
  // eslint-disable-next-line no-underscore-dangle
  return isContentItem(x) && x._content_type === "search_result";
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  return ["content_title"];
};

export type UrlType =
  | "retail-sales-locations"
  | "proposed-sales-locations"
  | "vote";

export type WithUrls<T> = T & { urls: Record<UrlType, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType?: UrlType
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  if (urlType) {
    item.url = initialItem.urls[urlType];
  }
  return item;
};

export function getLikelyQueryParams(): string | undefined {
  const searchParams = parseQuery(
    window.location.search,
    window.decodeURIComponent
  );
  const { query, name, address, latitude, longitude } = searchParams;
  const latlon =
    latitude && longitude ? `${latitude}, ${longitude}` : undefined;
  return query || name || address || latlon;
}

export const customRenderer = {
  createElement,
  Fragment,
  render
} as AutocompleteRenderer;
