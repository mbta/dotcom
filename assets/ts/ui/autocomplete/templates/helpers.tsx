/* eslint-disable import/prefer-default-export */
/* eslint-disable no-underscore-dangle */
import { SourceTemplates } from "@algolia/autocomplete-js";
import React from "react";
import { AutocompleteItem } from "../__autocomplete";
import { isAlgoliaItem, isContentItem, isSearchResultItem } from "../helpers";
import { LinkForItem } from "./algolia";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ItemTemplate = SourceTemplates<any>["item"];

export const itemURL = (item: AutocompleteItem): string => {
  let url = isContentItem(item) ? item._content_url : item.url;

  // Search result items are a subset of content items that point to a different URL
  if (isSearchResultItem(item)) {
    url = item._search_result_url.replace(/(internal|entity):/g, "/");
  }

  // Strip extra forward slashes as they break relative links
  url = url.replace(/\/\//g, "/");

  return url;
};

export const templateWithLink = (
  childTemplate: ItemTemplate
  // eslint-disable-next-line react/display-name
): ItemTemplate => props => {
  if (isAlgoliaItem(props.item)) {
    return (
      <LinkForItem
        item={props.item as AutocompleteItem}
        query={props.state.query}
      >
        {childTemplate(props)}
      </LinkForItem>
    );
  }
  return (
    <a href={props.item.url} className="aa-ItemLink">
      {childTemplate(props)}
    </a>
  );
};
