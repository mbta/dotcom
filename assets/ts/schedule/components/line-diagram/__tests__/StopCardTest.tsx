import { mount } from "enzyme";
import React from "react";
import * as redux from "react-redux";
import * as StopTreeHelpers from "../../../../helpers/stop-tree";
import { aroundNow } from "../../../../models/__tests__/alert-test";
import { Alert, InformedEntitySet, Route } from "../../../../__v3api";
import { RouteStop, StopTree } from "../../__schedule";
import { createStopTreeCoordStore } from "../graphics/useTreeStopPositions";
import StopCard from "../StopCard";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { testRouteStopListFromStopTree } from "../../../../app/helpers/testUtils";

const stopTree: StopTree = {
  byId: {
    a: {
      id: "a",
      value: ({
        id: "a",
        connections: [{ id: "Orange", name: "Orange Line" }],
        stop_features: ["parking_lot"]
      } as unknown) as RouteStop
    },
    b: {
      id: "b",
      value: ({
        id: "b",
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    c: {
      id: "c",
      value: ({
        id: "c",
        connections: [],
        stop_features: []
      } as unknown) as RouteStop
    },
    d: {
      id: "d",
      value: ({
        id: "d",
        connections: [],
        stop_features: [],
        branch: {},
        name: "Test",
        route: { id: "d", name: "Test" }
      } as unknown) as RouteStop
    },
    e: {
      id: "e",
      value: ({
        id: "e",
        connections: [{ id: "Green-Test" }],
        stop_features: [],
        branch: {},
        name: "Should Not Display This String",
        route: { id: "Green-E", name: "Green-E", type: 0 }
      } as unknown) as RouteStop
    },
    f: {
      id: "f",
      value: ({
        id: "f",
        connections: [],
        stop_features: [],
        branch: {},
        name: "Commuter Test",
        route: { id: "f", name: "Commuter Test", type: 2 }
      } as unknown) as RouteStop
    }
  },
  edges: {
    a: { next: ["b"], previous: [] },
    b: { next: ["c", "d", "e", "f"], previous: ["a"] },
    c: { next: [], previous: ["b"] },
    d: { next: [], previous: ["b"] },
    e: { next: [], previous: ["b"] },
    f: { next: [], previous: ["b"] }
  },
  startingNodes: ["a"]
};
const store = createStopTreeCoordStore(stopTree);
const testRouteStopList = testRouteStopListFromStopTree(stopTree);

const alertA: Alert = {
  id: "MOCK-ALERT-A",
  severity: 7,
  priority: "high",
  lifecycle: "new",
  effect: "detour",
  informed_entity: { stop: ["a"] } as InformedEntitySet,
  active_period: aroundNow()
} as Alert;
const alertB: Alert = {
  id: "MOCK-ALERT-B",
  severity: 7,
  priority: "high",
  lifecycle: "new",
  effect: "detour",
  informed_entity: { stop: ["b"] } as InformedEntitySet,
  active_period: aroundNow()
} as Alert;
const alertC: Alert = {
  id: "MOCK-ALERT-C",
  severity: 7,
  priority: "high",
  lifecycle: "upcoming",
  effect: "detour",
  informed_entity: { stop: ["a"] } as InformedEntitySet,
  active_period: aroundNow()
} as Alert;

const handleStopClick = () => {};

describe("StopCard", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders and matches snapshot", () => {
    const wrapper = mount(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[alertA, alertB]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("shows a View departures button", () => {
    jest.spyOn(StopTreeHelpers, "stopForId").mockImplementation(
      () =>
        (({
          route: { type: 3 } as Route,
          stop_features: [],
          connections: []
        } as unknown) as RouteStop)
    );

    const wrapper = mount(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );

    expect(wrapper.exists(".m-schedule-diagram__footer > button")).toBeTruthy();
    expect(
      wrapper.find(".m-schedule-diagram__footer > button").text()
    ).toContain("View departures");
  });

  it("should show high priority or high severity alerts", () => {
    const wrapperWithAlerts = mount(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[alertA, alertB]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );
    const alerts = wrapperWithAlerts.find(
      ".m-schedule-diagram__stop-link .c-svg__icon-alerts-triangle"
    );
    expect(alerts.length).toEqual(1);
  });

  it("should not show upcoming alerts", () => {
    const { container } = render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[alertC]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );

    expect(container.querySelector(".c-svg__icon-alerts-triangle")).toBeNull();
  });

  it("should show tooltip content", () => {
    const wrapper = mount(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[alertA, alertB]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );

    const connections = wrapper.find(".m-schedule-diagram__connections");

    const names = connections
      .find("a")
      .map(c => c.getDOMNode<HTMLElement>().dataset.originalTitle);
    expect(names).toEqual(["Orange Line"]);

    const features = wrapper.find(".m-schedule-diagram__features");

    const featureNames = features
      .find("span[data-toggle='tooltip']")
      .map(c => c.getDOMNode<HTMLElement>().dataset.originalTitle);
    expect(featureNames).toEqual(["Parking"]);
  });

  it("indicates detours, stop closures, etc", () => {
    const wrapper = mount(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"a"}
          alerts={[alertA, alertB]}
          onClick={handleStopClick}
        />
      </redux.Provider>
    );

    expect(wrapper.exists(".m-schedule-diagram__alert")).toBeTruthy();
    expect(wrapper.find(".m-schedule-diagram__alert").text()).toContain(
      "Detour"
    );
  });

  it("should display the line name on the card if there is a branch", () => {
    render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"d"}
          alerts={[]}
          onClick={() => {}}
        />
      </redux.Provider>
    );

    expect(screen.queryByText("Test Branch")).not.toBeNull();
  });

  it("should display the correct green line name if there is a branch", () => {
    render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"e"}
          alerts={[]}
          onClick={() => {}}
        />
      </redux.Provider>
    );

    expect(screen.queryByText("Green Line E Branch")).not.toBeNull();
  });

  it("should display the line name on the card if there is a branch and it is a commuter rail", () => {
    render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"f"}
          alerts={[]}
          onClick={() => {}}
        />
      </redux.Provider>
    );

    expect(screen.queryByText("Commuter Test Line")).not.toBeNull();
  });

  it("should display the green line connections from the route", () => {
    const { container } = render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"e"}
          alerts={[]}
          onClick={() => {}}
        />
      </redux.Provider>
    );

    expect(
      container.querySelectorAll("[href='/schedules/Green-Test/line']").length
    ).toBe(1);
  });

  it("should call the onClick function when the user clicks on the footer button", async () => {
    const clickSpy = jest.fn();
    render(
      <redux.Provider store={store}>
        <StopCard
          stopTree={stopTree}
          routeStopList={testRouteStopList}
          stopId={"b"}
          alerts={[]}
          onClick={clickSpy}
        />
      </redux.Provider>
    );

    const buttons = screen.getAllByRole("button");
    await userEvent.click(buttons[buttons.length - 1]);

    await waitFor(() => {
      expect(clickSpy).toHaveBeenCalled();
    });
  });
});
