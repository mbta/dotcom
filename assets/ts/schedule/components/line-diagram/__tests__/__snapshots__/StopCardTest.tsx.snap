// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopCard renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <StopCard stopTree={{...}} routeStopList={{...}} stopId=\\"a\\" alerts={{...}} onClick={[Function: handleStopClick]} liveData={{...}}>
    <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
      <section className=\\"m-schedule-diagram__content\\">
        <header className=\\"m-schedule-diagram__stop-heading\\">
          <h4 className=\\"m-schedule-diagram__stop-link notranslate\\">
            <a href=\\"/stops/a\\">
              <Alert>
                <span className=\\"notranslate c-svg__icon-alerts-triangle\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                <span className=\\"sr-only\\">
                  Service alert or delay
                </span>
                Â 
              </Alert>
              <MatchHighlight text={[undefined]} matchQuery={[undefined]}>
                <span />
              </MatchHighlight>
            </a>
          </h4>
          <div className=\\"m-schedule-diagram__features\\">
            <TooltipWrapper tooltipText=\\"Parking\\" tooltipOptions={{...}}>
              <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Parking\\">
                <i className=\\"fa fa-square-parking notranslate u-color-gray-light m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" />
              </span>
            </TooltipWrapper>
          </div>
        </header>
        <div className=\\"m-schedule-diagram__stop-details\\">
          <StopConnections connections={{...}}>
            <div className=\\"m-schedule-diagram__connections\\">
              <TooltipWrapper href=\\"/schedules/Orange/line\\" tooltipText=\\"Orange Line\\" tooltipOptions={{...}}>
                <a href=\\"/schedules/Orange/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Orange Line\\">
                  <span className=\\"m-schedule-diagram__connection\\">
                    <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  </span>
                </a>
              </TooltipWrapper>
            </div>
          </StopConnections>
          <div className=\\"m-schedule-diagram__alert\\">
            Detour
          </div>
        </div>
        <footer className=\\"m-schedule-diagram__footer\\">
          <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
            View schedule
          </button>
        </footer>
      </section>
    </li>
  </StopCard>
</Provider>"
`;
